### PR TITLE
Added quark parser to `ResourceParser`

### LINKED ISSUE #346: Add `Quark` support to scoop up any missed files within a Maven project.
See https://github.com/openrewrite/rewrite/pull/1737

We need to add the QuarkParser as the last parser in the resource parser. This will pick up any misc. files that have been missed by  Rewrite's other parsers. Additionally, by putting this in the ResourceParse, any files within the main/test folders will have the proper `SourceSet` assigned to them.

### PR DESCRIPTION
**Pull request by traceyyoshima**
Changes:

- Files ignored due to threshold are parsed as a `Quark`.
- Added `QuarkParser` to `ResourceParser`.
- ResourceParser will only parse files in the current maven project.
- Replaced `Files.find` with `Files.walkFileTree`.

fixes #346

### CODE REVIEW

File: src/main/java/org/openrewrite/maven/ResourceParser.java [hunk 1]
```diff
@@ -106,4 +111,55 @@ public <S extends SourceFile> List<S> parseSourceFiles(
             throw new UncheckedIOException(e);
         }
     }
+
+    public <S extends SourceFile> List<S> parseQuarks(
```
Comment (tkvangorder) [inline src/main/java/org/openrewrite/maven/ResourceParser.java:115]: I believe there is a static method on the QuarkParser that you can use rather than rolling your own. 

The only issue here is that the static method does not have the ability to pass a separate `baseDir` and `searchDir` and it does not allow an ExecutionContext to be passed to it. 

I would just overload the static method in QuarkParser to add these two parameters. 

This method becomes much simplier: 

```
List<Quark> quarks = QuarkParser.parseAllOtherFiles(baseDir, searchDir, alreadyParsed, ctx);
alreadyParsed.addAll(quarks.stream.map(SourceFile::getPath).collect(Collectors.toSet()));
return (List<S>) quarks;
```

**Review by tkvangorder — CHANGES_REQUESTED** [review]
Let me know if you have any questions.

File: src/main/java/org/openrewrite/maven/ResourceParser.java [hunk 2]
```diff
@@ -90,8 +92,7 @@ public <S extends SourceFile> List<S> parseSourceFiles(
 
             long fileSize = attrs.size();
             if ((sizeThresholdMb > 0 && fileSize > sizeThresholdMb * 1024L * 1024L)) {
-                alreadyParsed.add(path);
-                logger.info("Skipping parsing " + path + " as its size + " + fileSize / (1024L * 1024L) +
+                logger.info("Parsing as Quark " + path + " as its size + " + fileSize / (1024L * 1024L) +
                         "Mb exceeds size threshold " + sizeThresholdMb + "Mb");
                 return false;
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/maven/ResourceParser.java:97]: You should parse it as a quark right here and add it to the `alreadyParsed` list or this message will be displayed n times, where "n" is the number of parsers that come before the quark parser.

File: src/main/java/org/openrewrite/maven/ResourceParser.java [hunk 3]
```diff
@@ -106,4 +107,56 @@ public <S extends SourceFile> List<S> parseSourceFiles(
...
+    public <S extends SourceFile> List<S> parseQuarks(
+            Path baseDir,
+            Path searchDir,
+            Collection<Path> alreadyParsed,
+            ExecutionContext ctx) {
+        QuarkParser parser = new QuarkParser();
+        try (Stream<Path> resources = Files.find(searchDir, 16, (path, attrs) -> {
+            if (path.toString().endsWith(".java")) {
+                return false;
+            }
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/maven/ResourceParser.java:120]: Why is this necessary?

**Review by sambsnyd — CHANGES_REQUESTED** [review]

File: src/main/java/org/openrewrite/maven/ResourceParser.java (hunk 2 above)
Comment (sambsnyd) [reply src/main/java/org/openrewrite/maven/ResourceParser.java:97]: Alternatively, refactoring `ResourceParser` to not re-traverse the search directory for every single parser could also address this.

File: src/main/java/org/openrewrite/maven/ResourceParser.java (hunk 3 above)
Comment (sambsnyd) [reply src/main/java/org/openrewrite/maven/ResourceParser.java:120]: Oh I see we run `ResourceParser` before `JavaParser`. Makes sense

Comment (traceyyoshima) [reply src/main/java/org/openrewrite/maven/ResourceParser.java:120]: I ran the maven plugin on a few projects and java files are added without the condition.

- `.java` files aren't added to already parsed. I tested using `listJavaSources` to add the paths as already parsed, but the files were still added as quarks.
- The other resource parsers only accept specific types of files `.yaml`, `.java`, `.xml`, but the `QuarkParser#accept()` method always returns `true`.

Comment (tkvangorder) [reply src/main/java/org/openrewrite/maven/ResourceParser.java]: I think a better, long-term solution, would just be to map the paths for all the Java source files into the `alreadyParsed` and not have any specialized logic. I envision a world where there will also be groovy and Kotlin parsers in the mix and I would not want to have separate logic to exclude certain file masks (like "*.java")

Comment (tkvangorder) [reply src/main/java/org/openrewrite/maven/ResourceParser.java]: We do parse the Java sources before the resources (I believe)

Comment (tkvangorder) [reply src/main/java/org/openrewrite/maven/ResourceParser.java]: Is there something specialized in `parseQuarks` that is not handled by `parseSourceFiles`? It seems largely a duplicate of that method.

File: src/main/java/org/openrewrite/maven/ResourceParser.java [hunk 4]
```diff
@@ -90,17 +87,77 @@ public <S extends SourceFile> List<S> parseSourceFiles(
 
             long fileSize = attrs.size();
             if ((sizeThresholdMb > 0 && fileSize > sizeThresholdMb * 1024L * 1024L)) {
-                alreadyParsed.add(path);
-                logger.info("Skipping parsing " + path + " as its size + " + fileSize / (1024L * 1024L) +
+                logger.info("Parsing as Quark " + path + " as its size + " + fileSize / (1024L * 1024L) +
                         "Mb exceeds size threshold " + sizeThresholdMb + "Mb");
-                return false;
+                return true;
             }
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/maven/ResourceParser.java:93]: Size threshold doesn't work anymore. You return `true` here but there's no logic below to make sure that gets parsed as a quark

**Review by sambsnyd — CHANGES_REQUESTED** [review]

File: src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java [hunk 5]
```diff
@@ -306,10 +306,16 @@ private static RawRepositories buildRawRepositories(@Nullable List<Repository> r
         return rawRepositories;
     }
 
+    private Set<Path> skipOtherMavenProjects() {
+        return mavenSession.getProjects().stream()
```
Comment (traceyyoshima) [inline src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java:310]: Note for posterity:
`mavenSession#getProjects()` returns the projects that are already sorted in reactor order.

```java
    @Override
    public void execute() throws MojoExecutionException {
        if (mavenSession.getCurrentProject == mavenSession.getProjects().get(mavenSession.getProjects().size() - 1)) {
           // do something when the last project is building.
        }
    }
```

File: src/main/java/org/openrewrite/maven/ResourceParser.java [hunk 6]
```diff
@@ -3,107 +3,168 @@
...
+            } else if (yamlParser.accept(path)) {
+                yamlPaths.add(path);
+            } else if (propertiesParser.accept(path)) {
+                propertiesPaths.add(path);
+            } else if (protoParser.accept(path)) {
+                protoPaths.add(path);
+            } else if (hclParser.accept(path)) {
+                hclPaths.add(path);
+            } else if (quarkParser.accept(path)) {
+                quarkPaths.add(path);
```
Comment (traceyyoshima) [inline src/main/java/org/openrewrite/maven/ResourceParser.java:130]: Java files that do not exist in a maven module or are in a module but exist outside of `main/test` are added as quarks. We could use the JavaParser to parse those as resources if preferred.

File: src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java [hunk 7]
```diff
@@ -342,11 +349,11 @@ public List<SourceFile> listSourceFiles(Iterable<NamedStyles> styles,
         sourceFiles.addAll(ListUtils.map(maybeAutodetectStyles(javaParser.parse(mainJavaSources, baseDir, ctx), styles),
                 addProvenance(baseDir, projectProvenance, generatedSourcePaths)));
 
-        ResourceParser rp = new ResourceParser(logger, exclusions, sizeThresholdMb);
+        ResourceParser rp = new ResourceParser(baseDir, logger, exclusions, sizeThresholdMb, skipOtherMavenProjects());
 
         // Any resources parsed from "main/resources" should also have the main source set added to them.
         sourceFiles.addAll(ListUtils.map(
-                rp.parse(baseDir, mavenProject.getBasedir().toPath().resolve("src/main/resources"), alreadyParsed),
+                rp.parse(mavenProject.getBasedir().toPath().resolve("src/main/resources"), alreadyParsed),
```
Comment (tkvangorder) [inline src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java:356]: I am a little concerned here because the `baseDir` is the root of the maven project and all paths are made relative to it....this helps when we are checking for exclusions...

To clarify here the instance variable `baseDir` should be the root directory for the entire maven project and `mavenProject.getBaseDir()` will be the directory of the current project.

I could be wrong here and maybe you are accounting for this below....just an observation.

File: src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java [hunk 8]
```diff
@@ -306,10 +306,16 @@ private static RawRepositories buildRawRepositories(@Nullable List<Repository> r
         return rawRepositories;
     }
 
+    private Set<Path> skipOtherMavenProjects() {
+        return mavenSession.getProjects().stream()
+                .filter(o -> o != mavenSession.getCurrentProject())
+                .map(o -> o.getBasedir().toPath())
```
Comment (tkvangorder) [inline src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java:312]: Same thing here, shouldn't these paths be resolved from `baseDir` (the instance variable)?

**Comment by traceyyoshima** [comment]
Closing the PR to separate changes from adding Quark support and #352.

File: src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java (hunk 7 above)
Comment (traceyyoshima) [reply src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java:356]: The `baseDir` is passed into the ResourceParser constructor on line 352.

Prior to the change, PathMatchers were generated using `baseDir.getFileSystem().getPathMatcher(..)` per path that is traversed. By moving the `baseDir` into the constructor, we don't need to pass the implied path into each method call. And the PathMatchers may be generated during instantiation of ResourceParser rather than doing the work repeatedly.

File: src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java (hunk 8 above)
Comment (traceyyoshima) [reply src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java:312]: No, I don't think so.
